EZEE-3303 - search engine comparison (#1176)
ezsystems/developer-documentation: docs/guide/search/search.md

eZ Platform exposes a very powerful [Search API](../../api/public_php_api_search.md), allowing both full-text search and querying the content Repository using several built-in Search Criteria and Sort Clauses. These are supported across different search engines, allowing you to plug in another search engine without changing your code.

Currently, the following search engines exist in their own eZ Platform Bundles:

1.  [Legacy search engine](search_engines.md#legacy-search-engine-bundle), a database-powered search engine for basic needs.
1.  [Solr](solr.md), an integration providing better overall performance, much better scalability and support for more advanced search capabilities.
1.  [Elasticsearch](elastic.md), available for eZ Enterprise customers, a document-oriented engine providing even better performance and scalability.

### Feature comparison

| Feature | Elasticsearch | Apache Solr | Legacy Search Engine (SQL) |
| --- | --- | --- | --- |
| Filtering | Yes | Yes | Yes, limited\* |
| Query (filter with scoring) | Yes | Yes | Only filters, no scoring |

> | Fulltext | Yes | Yes, incl. advanced features | Partial support |

Reviewer: ```suggestion
| Fulltext search | Yes | Yes, incl. advanced features | Partial support |
```

eZ Platform exposes a very powerful [Search API](../../api/public_php_api_search.md), allowing both full-text search and querying the content Repository using several built-in Search Criteria and Sort Clauses. These are supported across different search engines, allowing you to plug in another search engine without changing your code.

Currently, the following search engines exist in their own eZ Platform Bundles:

1.  [Legacy search engine](search_engines.md#legacy-search-engine-bundle), a database-powered search engine for basic needs.
1.  [Solr](solr.md), an integration providing better overall performance, much better scalability and support for more advanced search capabilities.
1.  [Elasticsearch](elastic.md), available for eZ Enterprise customers, a document-oriented engine providing even better performance and scalability.

### Feature comparison

> | **Feature** | **Elasticsearch** | **Apache Solr** | **Legacy Search Engine** _**(SQL)**_ |

Reviewer: ```suggestion
| Feature | Elasticsearch | Apache Solr | Legacy Search Engine (SQL) |
```

| --- | --- | --- | --- |

> | Filtering | Yes | Yes | Yes, but limited\* |

Reviewer: ```suggestion
| Filtering | Yes | Yes | Yes, limited\* |
```

> | Query _(filter with scoring)_ | Yes | Yes | Only filters, no scoring |

Reviewer: ```suggestion
| Query (filter with scoring) | Yes | Yes | Only filters, no scoring |
```

> | Fulltext | Yes | Yes, incl. advanced features | Partly\*\* |

Reviewer: ```suggestion
| Fulltext | Yes | Yes, incl. advanced features | Partial support \*\* |
```

> | Faceting | Partly (deprecated) | [Partly as of v1.4](https://doc.ezplatform.com/en/latest/api/public_php_api_browsing/#performing-a-faceted-search) | No |

Reviewer: ```suggestion
| Faceting (deprecated)  | Partial support | Partial support | No |
```
The oldest supported version uses the solr bundle > v1.4, so this is irrelevant.
Also remember to remove the "deprecated" note for faceting when merging to other versions (it is only deprecated in v3.2).

> | Spellchecking | Planned _(TBD when)_ | Planned _(TBD when)_ | No |
> | Highlighting | Planned _(TBD when)_ | Planned _(TBD when)_ | No |

Reviewer: ```suggestion
| Spellchecking | No | No | No |
| Highlighting | No | No | No |
```
Let's deal with what is in the product now, not in the future. Or you could just remove those two lines altogether

> | Index time boosting | ?\*\*\*\* | [Yes, in \&gt;= v1.4,\&lt; v2.0](https://github.com/ezsystems/ezplatform-solr-search-engine/releases/tag/v1.4.0)\*\*\* | No |

Reviewer: ```suggestion
| Index time boosting | \*\*\*\* | Yes | No |
```
AFAIK this means it is supported in v3, but not in earlier versions. Please indicate it when merging to relevant branches.

| Aggregations | Yes | Yes | No |

> _* Usage of Criterion and SortClause for Fields does not perform well on medium to larger 

Reviewer: ```suggestion
* Usage of Criteria and Sort Clauses for Fields does not perform well on medium to larger 
```

> amount of data with Legacy Search Engine (SQL), use Solr for this._

Reviewer: ```suggestion
amounts of data with Legacy Search Engine (SQL), use Solr for this.
```

> _** Does not include full set of full text features includes: Scoring/Ranking (Solr performs 
> sorting on scoring by default, for location search available as of [v1.3][3] with Solr 6.x), 
> and as of [Solr Search Engine v1.5](https://github.com/ezsystems/ezplatform-solr-search-engine/releases/tag/v1.5.0) supports advanced 
> full text capabilities such as `word "phrase" (group) +mandatory -prohibited AND && OR || NOT !`._

Reviewer: ```suggestion
** Does not cover all full text features including: scoring/ranking (Solr performs sorting on scoring by default)
```
Please double check, but I think all the warnings in this point referred to versions of the solr bundle that are not used in any supported version of Platform (v1.13 uses solr bundle v1.5)

> _*** As of Solr 7, index-time boosting is no longer available. Query time boosting is planned to be introduced instead in future 2.x and 3.x release.

Reviewer: ```suggestion
*** As of Solr 7, index-time boosting is no longer available.
```

> See [Major changes in Solr 7](https://lucene.apache.org/solr/guide/7_2/major-changes-in-solr-7.html) for more information._

Reviewer: ```suggestion
See [Major changes in Solr 7](https://lucene.apache.org/solr/guide/7_2/major-changes-in-solr-7.html) for more information.
```

> _**** Elasticsearch offers query time boosting instead._

Reviewer: ```suggestion
**** Elasticsearch offers query time boosting instead.
```

| Highlighting | Planned _(TBD when)_ | Planned _(TBD when)_ | No |
| Index time boosting | ?\*\*\*\* | [Yes, in \&gt;= v1.4,\&lt; v2.0](https://github.com/ezsystems/ezplatform-solr-search-engine/releases/tag/v1.4.0)\*\*\* | No |
| Aggregations | Yes | Yes | No |

_* Usage of Criterion and SortClause for Fields does not perform well on medium to larger 
amount of data with Legacy Search Engine (SQL), use Solr for this._

_** Does not include full set of full text features includes: Scoring/Ranking (Solr performs 
sorting on scoring by default, for location search available as of [v1.3][3] with Solr 6.x), 
and as of [Solr Search Engine v1.5](https://github.com/ezsystems/ezplatform-solr-search-engine/releases/tag/v1.5.0) supports advanced 
full text capabilities such as `word "phrase" (group) +mandatory -prohibited AND && OR || NOT !`._

_*** As of Solr 7, index-time boosting is no longer available. Query time boosting is planned to be introduced instead in future 2.x and 3.x release.
See [Major changes in Solr 7](https://lucene.apache.org/solr/guide/7_2/major-changes-in-solr-7.html) for more information._

_**** Elasticsearch offers query time boosting instead._

> ## Install

Author: ```suggestion
```
